src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkBaseExternalContext.java

```diff
@@ -1150,29 +1160,152 @@ protected StarlarkPath getPath(String method, Object path)
+    if (path.startsWith(directories.getWorkspace())) {
+      // The file is under the workspace root.
+      PathFragment relPath = path.relativeTo(directories.getWorkspace());
+      return Pair.of(
+          new FileInsideWorkspace(RepositoryName.MAIN, relPath),
+          RootedPath.toRootedPath(Root.fromPath(directories.getWorkspace()), relPath));
+    }
+    Path outputBaseExternal =
+        directories.getOutputBase().getRelative(LabelConstants.EXTERNAL_REPOSITORY_LOCATION);
+    if (path.startsWith(outputBaseExternal)) {
+      PathFragment relPath = path.relativeTo(outputBaseExternal);
+      if (!relPath.isEmpty()) {
+        // The file is under a repo root.
+        String repoName = relPath.getSegment(0);
+        PathFragment repoRelPath =
+            relPath.relativeTo(PathFragment.createAlreadyNormalized(repoName));
+        return Pair.of(
+            new FileInsideWorkspace(RepositoryName.createUnvalidated(repoName), repoRelPath),
+            RootedPath.toRootedPath(
+                Root.fromPath(outputBaseExternal.getChild(repoName)), repoRelPath));
+      }
+    }
+    // The file is just under a random absolute path.
+    return Pair.of(
+        new FileOutsideWorkspace(path.asFragment()),
+        RootedPath.toRootedPath(Root.absoluteRoot(path.getFileSystem()), path));
+  }
+
+  protected enum ShouldWatch {
+    YES,
+    NO,
+    AUTO;
+
+    static ShouldWatch fromString(String s) throws EvalException {
+      switch (s) {
+        case "yes":
+          return YES;
+        case "no":
+          return NO;
+        case "auto":
+          return AUTO;
+      }
+      throw Starlark.errorf(
+          "bad value for 'watch' parameter; want 'yes', 'no', or 'auto', got %s", s);
+    }
+  }
+
+  protected void maybeWatch(StarlarkPath starlarkPath, ShouldWatch shouldWatch)
+      throws EvalException, RepositoryFunctionException, InterruptedException {
+    if (shouldWatch == ShouldWatch.NO) {
+      return;
+    }
+    Path path = starlarkPath.getPath();
+    Pair<RepoRecordedInput.File, RootedPath> pair = toRootedPath(path);
+    if (path.startsWith(workingDirectory)) {
+      // The file is under the working directory. Don't watch it, as it would cause a dependency
+      // cycle.
+      if (shouldWatch == ShouldWatch.AUTO) {
+        return;
+      }
+      throw Starlark.errorf("attempted to watch file under working directory");
+    }
+    if (!allowWatchingFilesOutsideWorkspace && pair.first instanceof FileOutsideWorkspace) {
+      if (shouldWatch == ShouldWatch.AUTO) {
+        return;
+      }
+      throw Starlark.errorf(
+          "attempted to watch file outside workspace, but it's prohibited in the current context");
+    }
+    try {
+      FileValue fileValue =
+          (FileValue) env.getValueOrThrow(FileValue.key(pair.second), IOException.class);
+      if (fileValue == null) {
+        throw new NeedsSkyframeRestartException();
```

Comment:
Does this mean that if you watch N files previously unseen, it's O(N^2) time and re-doing the rest of the work the repository does N times?

Reply:
This is only O(N^2) if we actually do restart skyframe; we're now using loom by default so it's actually linear. Given Loom will be the way forward, I don't want to spend more engineering time for the no-Loom case (i.e. the one-single-getValues()-call-at-the-end thing); in fact I want to remove more workarounds for repo restarts (e.g. [`enforceLabelAttributes`](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkRepositoryContext.java;drc=e8ac96adb860b9831a62681167295f103a51470a;l=519))

Reply:
> it turns out that that use case went away a few hours ago

just out of curiosity, what is that use case?

Reply:
The builds of the Android kernel. I got word overnight that they finished their upgrade to JVM 21.